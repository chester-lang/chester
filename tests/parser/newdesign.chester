module ğŸ˜¿ğŸ˜¿;

trait #sealed Vector[n: Nat, T: Type];

data Nil <: Vector[0, T];

data Cons[n, T] <: Vector[succ(n), T] {
  head: T;
  tail: Vector[n, T];
}

cons1 = Cons(0, Nil);
cons2 = Cons %{ head = 0, tail = Nil };

implement Vector[n, T] {
  isEmpty: Boolean = ?todo;
}

module Vector {
  empty[T]: Vector[0, T] = Nil;
}