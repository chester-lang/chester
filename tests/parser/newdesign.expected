Block(
  statements = Vector(
    OpSeq(seq = Vector(Identifier(name = "module", meta = None), Identifier(name = "ðŸ˜¿ðŸ˜¿", meta = None)), meta = None),
    OpSeq(
      seq = Vector(
        Identifier(name = "trait", meta = None),
        Keyword(key = "sealed", telescope = Vector(), meta = None),
        FunctionCall(
          function = Identifier(name = "Vector", meta = None),
          telescope = ListExpr(
            terms = Vector(
              OpSeq(
                seq = Vector(
                  Identifier(name = "n", meta = None),
                  Identifier(name = ":", meta = None),
                  Identifier(name = "Nat", meta = None)
                ),
                meta = None
              ),
              OpSeq(
                seq = Vector(
                  Identifier(name = "T", meta = None),
                  Identifier(name = ":", meta = None),
                  Identifier(name = "Type", meta = None)
                ),
                meta = None
              )
            ),
            meta = None
          ),
          meta = None
        )
      ),
      meta = None
    ),
    OpSeq(
      seq = Vector(
        Identifier(name = "data", meta = None),
        Identifier(name = "Nil", meta = None),
        Identifier(name = "<:", meta = None),
        FunctionCall(
          function = Identifier(name = "Vector", meta = None),
          telescope = ListExpr(
            terms = Vector(IntegerLiteral(value = 0, meta = None), Identifier(name = "T", meta = None)),
            meta = None
          ),
          meta = None
        )
      ),
      meta = None
    ),
    OpSeq(
      seq = Vector(
        Identifier(name = "data", meta = None),
        FunctionCall(
          function = Identifier(name = "Cons", meta = None),
          telescope = ListExpr(
            terms = Vector(Identifier(name = "n", meta = None), Identifier(name = "T", meta = None)),
            meta = None
          ),
          meta = None
        ),
        Identifier(name = "<:", meta = None),
        FunctionCall(
          function = Identifier(name = "Vector", meta = None),
          telescope = ListExpr(
            terms = Vector(
              FunctionCall(
                function = Identifier(name = "succ", meta = None),
                telescope = Tuple(terms = Vector(Identifier(name = "n", meta = None)), meta = None),
                meta = None
              ),
              Identifier(name = "T", meta = None)
            ),
            meta = None
          ),
          meta = None
        ),
        Block(
          statements = Vector(
            OpSeq(
              seq = Vector(
                Identifier(name = "head", meta = None),
                Identifier(name = ":", meta = None),
                Identifier(name = "T", meta = None)
              ),
              meta = None
            ),
            OpSeq(
              seq = Vector(
                Identifier(name = "tail", meta = None),
                Identifier(name = ":", meta = None),
                FunctionCall(
                  function = Identifier(name = "Vector", meta = None),
                  telescope = ListExpr(
                    terms = Vector(Identifier(name = "n", meta = None), Identifier(name = "T", meta = None)),
                    meta = None
                  ),
                  meta = None
                )
              ),
              meta = None
            )
          ),
          result = None,
          meta = None
        )
      ),
      meta = None
    ),
    OpSeq(
      seq = Vector(
        Identifier(name = "cons1", meta = None),
        Identifier(name = "=", meta = None),
        FunctionCall(
          function = Identifier(name = "Cons", meta = None),
          telescope = Tuple(
            terms = Vector(IntegerLiteral(value = 0, meta = None), Identifier(name = "Nil", meta = None)),
            meta = None
          ),
          meta = None
        )
      ),
      meta = None
    ),
    OpSeq(
      seq = Vector(
        Identifier(name = "cons2", meta = None),
        Identifier(name = "=", meta = None),
        Identifier(name = "Cons", meta = None),
        ObjectExpr(
          clauses = Vector(
            ObjectExprClause(key = Identifier(name = "head", meta = None), value = IntegerLiteral(value = 0, meta = None)),
            ObjectExprClause(key = Identifier(name = "tail", meta = None), value = Identifier(name = "Nil", meta = None))
          ),
          meta = None
        )
      ),
      meta = None
    ),
    OpSeq(
      seq = Vector(
        Identifier(name = "implement", meta = None),
        FunctionCall(
          function = Identifier(name = "Vector", meta = None),
          telescope = ListExpr(
            terms = Vector(Identifier(name = "n", meta = None), Identifier(name = "T", meta = None)),
            meta = None
          ),
          meta = None
        ),
        Block(
          statements = Vector(
            OpSeq(
              seq = Vector(
                Identifier(name = "isEmpty", meta = None),
                Identifier(name = ":", meta = None),
                Identifier(name = "Boolean", meta = None),
                Identifier(name = "=", meta = None),
                Identifier(name = "?", meta = None),
                Identifier(name = "todo", meta = None)
              ),
              meta = None
            )
          ),
          result = None,
          meta = None
        )
      ),
      meta = None
    ),
    OpSeq(
      seq = Vector(
        Identifier(name = "module", meta = None),
        Identifier(name = "Vector", meta = None),
        Block(
          statements = Vector(
            OpSeq(
              seq = Vector(
                FunctionCall(
                  function = Identifier(name = "empty", meta = None),
                  telescope = ListExpr(terms = Vector(Identifier(name = "T", meta = None)), meta = None),
                  meta = None
                ),
                Identifier(name = ":", meta = None),
                FunctionCall(
                  function = Identifier(name = "Vector", meta = None),
                  telescope = ListExpr(
                    terms = Vector(IntegerLiteral(value = 0, meta = None), Identifier(name = "T", meta = None)),
                    meta = None
                  ),
                  meta = None
                ),
                Identifier(name = "=", meta = None),
                Identifier(name = "Nil", meta = None)
              ),
              meta = None
            )
          ),
          result = None,
          meta = None
        )
      ),
      meta = None
    )
  ),
  result = None,
  meta = None
)